ENIGMAsystem/SHELL/Widget_Systems/KDialog/dialogs.cpp

```diff
@@ -15,10 +15,523 @@
+  string("--attach=$(xprop -root 32x '\t$0' _NET_ACTIVE_WINDOW | cut -f 2) ") +
+  string("--password \"") + add_escaping(str, false, "") + string("\" \"") +
+  add_escaping(def, false, "") + string("\" --title \"") +
+  str_title + string("\");echo $ans");
+
+  return shellscript_evaluate(str_command);
+}
+
+double get_integer(string str, double def) {
+  string str_def = remove_trailing_zeros(def);
+  string str_result = get_string(str, str_def);
+  return strtod(str_result.c_str(), NULL);
+}
+
+double get_passcode(string str, double def) {
+  string str_def = remove_trailing_zeros(def);
+  string str_result = get_password(str, str_def);
+  return strtod(str_result.c_str(), NULL);
+}
+
+string get_open_filename(string filter, string fname) {
+  string str_command; string pwd;
+  string str_title = "Open";
+  string str_fname = filename_name(fname);
+
+  pwd = ""; if (str_fname.c_str() && str_fname[0] != '/' && str_fname.length()) pwd = string("\"$PWD/\"") +
+    string("\"") + add_escaping(str_fname, false, "") + string("\""); else pwd = "\"$PWD/\"";
+
+  str_command = string("ans=$(kdialog ") +
+  string("--attach=$(xprop -root 32x '\t$0' _NET_ACTIVE_WINDOW | cut -f 2) ") +
+  string("--getopenfilename /") + pwd + add_escaping(kdialog_filter(filter), false, "") +
+  string(" --title \"") + str_title + string("\"") + string(");echo $ans");
+
+  string result = shellscript_evaluate(str_command);
+  return file_exists(result) ? result : "";
+}
+
+string get_open_filename_ext(string filter, string fname, string dir, string title) {
+  string str_command; string pwd;
+  string str_title = add_escaping(title, true, "Open");
+  string str_fname = filename_name(fname);
+  string str_dir = filename_path(dir);
+
+  string str_path = fname;
+  if (str_dir[0] != '\0') str_path = str_dir + str_fname;
+  str_fname = (char *)str_path.c_str();
+
+  pwd = ""; if (str_fname.c_str() && str_fname[0] != '/' && str_fname.length()) pwd = string("\"$PWD/\"") +
+    string("\"") + add_escaping(str_fname, false, "") + string("\""); else pwd = "\"$PWD/\"";
+
+  str_command = string("ans=$(kdialog ") +
+  string("--attach=$(xprop -root 32x '\t$0' _NET_ACTIVE_WINDOW | cut -f 2) ") +
+  string("--getopenfilename /") + pwd + add_escaping(kdialog_filter(filter), false, "") +
+  string(" --title \"") + str_title + string("\"") + string(");echo $ans");
+
+  string result = shellscript_evaluate(str_command);
+  return file_exists(result) ? result : "";
+}
+
+string get_open_filenames(string filter, string fname) {
+  string str_command; string pwd;
+  string str_title = "Open";
+  string str_fname = filename_name(fname);
+
+  pwd = ""; if (str_fname.c_str() && str_fname[0] != '/' && str_fname.length()) pwd = string("\"$PWD/\"") +
+    string("\"") + add_escaping(str_fname, false, "") + string("\""); else pwd = "\"$PWD/\"";
+
+  str_command = string("kdialog ") +
+  string("--attach=$(xprop -root 32x '\t$0' _NET_ACTIVE_WINDOW | cut -f 2) ") +
```

Comment:
There's no need for `string("") + string("")`; just use `string(""  "")`. In fact, the `string()` construction is unnecessary, here, anyway, since `pwd` is a string.